File: starters/dotnet/FProto.Core/CryptoUtils.cs
using System.Buffers.Binary;
using System.Security.Cryptography;

namespace FProto.Core;

public static class CryptoUtils
{
    public static byte[] BuildNonce(ulong seq)
    {
        var nonce = new byte[12];
        BinaryPrimitives.WriteUInt64LittleEndian(nonce.AsSpan(4), seq);
        return nonce;
    }

    public static byte[] BuildAad(byte[] sessionId, ulong seq)
    {
        var aad = new byte[sessionId.Length + 8];
        sessionId.CopyTo(aad, 0);
        BinaryPrimitives.WriteUInt64LittleEndian(aad.AsSpan(sessionId.Length), seq);
        return aad;
    }

    public static byte[] EncryptAead(byte[] key, byte[] sessionId, ulong seq, byte[] plaintext)
    {
        if (key.Length != 32)
            throw new ArgumentException($"invalid key size: expected 32, got {key.Length}");

        var nonce = BuildNonce(seq);
        var aad = BuildAad(sessionId, seq);

        using var cipher = new ChaCha20Poly1305(key);
        var ciphertext = new byte[plaintext.Length];
        var tag = new byte[16];
        cipher.Encrypt(nonce, plaintext, ciphertext, tag, aad);

        var result = new byte[ciphertext.Length + tag.Length];
        ciphertext.CopyTo(result, 0);
        tag.CopyTo(result, ciphertext.Length);
        return result;
    }

    public static byte[] DecryptAead(byte[] key, byte[] sessionId, ulong seq, byte[] ciphertextWithTag)
    {
        if (key.Length != 32)
            throw new ArgumentException($"invalid key size: expected 32, got {key.Length}");

        var nonce = BuildNonce(seq);
        var aad = BuildAad(sessionId, seq);

        var ciphertext = ciphertextWithTag.AsSpan(0, ciphertextWithTag.Length - 16);
        var tag = ciphertextWithTag.AsSpan(ciphertextWithTag.Length - 16);

        using var cipher = new ChaCha20Poly1305(key);
        var plaintext = new byte[ciphertext.Length];
        cipher.Decrypt(nonce, ciphertext, tag, plaintext, aad);
        return plaintext;
    }

    public static byte[] Rekey(byte[] currentKey)
    {
        var prk = HMACSHA256.HashData(Array.Empty<byte>(), currentKey);
        var info = "fproto-rekey"u8.ToArray();
        var input = new byte[info.Length + 1];
        info.CopyTo(input, 0);
        input[^1] = 0x01;
        var okm = HMACSHA256.HashData(prk, input);
        return okm[..32];
    }

    public static void Zeroize(byte[] buf)
    {
        CryptographicOperations.ZeroMemory(buf);
    }
}

File: starters/java/src/main/java/io/fproto/core/CryptoUtils.java
package io.fproto.core;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.util.Arrays;

public final class CryptoUtils {

    private CryptoUtils() {}

    public static KeyPair generateKeyPair() throws Exception {
        KeyPairGenerator kpg = KeyPairGenerator.getInstance("X25519");
        return kpg.generateKeyPair();
    }

    public static byte[] buildNonce(long seq) {
        byte[] nonce = new byte[12];
        ByteBuffer.wrap(nonce, 4, 8).order(ByteOrder.LITTLE_ENDIAN).putLong(seq);
        return nonce;
    }

    public static byte[] buildAad(byte[] sessionId, long seq) {
        byte[] aad = new byte[sessionId.length + 8];
        System.arraycopy(sessionId, 0, aad, 0, sessionId.length);
        ByteBuffer.wrap(aad, sessionId.length, 8).order(ByteOrder.LITTLE_ENDIAN).putLong(seq);
        return aad;
    }

    public static byte[] encryptAead(byte[] key, byte[] sessionId, long seq, byte[] plaintext) throws Exception {
        if (key.length != 32) {
            throw new IllegalArgumentException("invalid key size: expected 32, got " + key.length);
        }
        byte[] nonce = buildNonce(seq);
        byte[] aad = buildAad(sessionId, seq);

        Cipher cipher = Cipher.getInstance("ChaCha20-Poly1305");
        SecretKeySpec keySpec = new SecretKeySpec(key, "ChaCha20");
        IvParameterSpec ivSpec = new IvParameterSpec(nonce);
        cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec);
        cipher.updateAAD(aad);
        return cipher.doFinal(plaintext);
    }

    public static byte[] decryptAead(byte[] key, byte[] sessionId, long seq, byte[] ciphertext) throws Exception {
        if (key.length != 32) {
            throw new IllegalArgumentException("invalid key size: expected 32, got " + key.length);
        }
        byte[] nonce = buildNonce(seq);
        byte[] aad = buildAad(sessionId, seq);

        Cipher cipher = Cipher.getInstance("ChaCha20-Poly1305");
        SecretKeySpec keySpec = new SecretKeySpec(key, "ChaCha20");
        IvParameterSpec ivSpec = new IvParameterSpec(nonce);
        cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec);
        cipher.updateAAD(aad);
        return cipher.doFinal(ciphertext);
    }

    public static byte[] rekey(byte[] currentKey) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(new byte[0], "HmacSHA256"));
        byte[] prk = mac.doFinal(currentKey);

        Mac mac2 = Mac.getInstance("HmacSHA256");
        mac2.init(new SecretKeySpec(prk, "HmacSHA256"));
        mac2.update("fproto-rekey".getBytes());
        mac2.update((byte) 0x01);
        byte[] okm = mac2.doFinal();
        return Arrays.copyOf(okm, 32);
    }

    public static void zeroize(byte[] buf) {
        Arrays.fill(buf, (byte) 0);
    }
}
